File: sdk/python/controldb/transport.py
"""HTTP transport with retry, batching, and an offline spool.

The transport is the only object that knows about the wire. The Run and Client
code build envelopes and hand them off here.
"""

from __future__ import annotations

import json
import os
import random
import threading
import time
from pathlib import Path
from typing import Any, Dict, Iterable, List, Mapping, Optional, Sequence

import httpx

from .errors import (
    AuthenticationError,
    AuthorizationError,
    CollectorUnavailableError,
    IdempotencyConflictError,
    PayloadTooLargeError,
    ValidationError,
)


def _raise_for_status(resp: httpx.Response) -> None:
    if resp.status_code < 400:
        return
    body: Any
    try:
        body = resp.json()
    except Exception:
        body = resp.text
    message = body.get("detail") if isinstance(body, dict) else str(body)
    if resp.status_code == 401:
        raise AuthenticationError(message or "unauthenticated")
    if resp.status_code == 403:
        raise AuthorizationError(message or "forbidden")
    if resp.status_code == 409:
        raise IdempotencyConflictError(message or "conflict")
    if resp.status_code == 413:
        raise PayloadTooLargeError(message or "payload too large")
    if resp.status_code == 422:
        raise ValidationError(message or "validation error")
    if resp.status_code >= 500:
        raise CollectorUnavailableError(message or f"collector status {resp.status_code}")
    raise CollectorUnavailableError(message or f"unexpected status {resp.status_code}")


class Spool:
    """Append-only on-disk spool for offline buffering."""

    def __init__(self, path: Path):
        self.path = path
        self.path.parent.mkdir(parents=True, exist_ok=True)
        self._lock = threading.Lock()

    def write(self, batch: Mapping[str, Any]) -> None:
        with self._lock, self.path.open("a", encoding="utf-8") as fh:
            fh.write(json.dumps(batch) + "\n")

    def drain(self) -> List[Dict[str, Any]]:
        with self._lock:
            if not self.path.exists():
                return []
            entries: List[Dict[str, Any]] = []
            with self.path.open("r", encoding="utf-8") as fh:
                for line in fh:
                    line = line.strip()
                    if not line:
                        continue
                    entries.append(json.loads(line))
            self.path.unlink()
            return entries


class Transport:
    def __init__(
        self,
        base_url: str,
        api_key: str,
        *,
        timeout: float = 10.0,
        max_retries: int = 4,
        backoff_initial: float = 0.2,
        backoff_max: float = 5.0,
        fail_open: bool = False,
        spool_path: Optional[str] = None,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.timeout = timeout
        self.max_retries = max_retries
        self.backoff_initial = backoff_initial
        self.backoff_max = backoff_max
        self.fail_open = fail_open
        self.spool = Spool(Path(spool_path)) if spool_path else None
        self._client = httpx.Client(
            base_url=self.base_url,
            timeout=timeout,
            headers={
                "Authorization": f"Bearer {api_key}",
                "User-Agent": "controldb-python/0.1",
                "Content-Type": "application/json",
            },
        )

    def close(self) -> None:
        try:
            self._client.close()
        except Exception:
            pass

    def request(
        self,
        method: str,
        path: str,
        *,
        json_body: Optional[Mapping[str, Any]] = None,
        idempotency_key: Optional[str] = None,
        retry: bool = True,
    ) -> Dict[str, Any]:
        headers: Dict[str, str] = {}
        if idempotency_key:
            headers["Idempotency-Key"] = idempotency_key
        attempts = self.max_retries if retry else 1
        last_error: Optional[Exception] = None
        for attempt in range(attempts):
            try:
                resp = self._client.request(method, path, json=json_body, headers=headers)
                _raise_for_status(resp)
                if resp.status_code == 204 or not resp.content:
                    return {}
                return resp.json()
            except (httpx.HTTPError, CollectorUnavailableError) as exc:
                last_error = exc
                if attempt + 1 >= attempts:
                    break
                sleep = min(self.backoff_max, self.backoff_initial * (2 ** attempt))
                sleep += random.uniform(0, sleep / 2)
                time.sleep(sleep)
            except (AuthenticationError, AuthorizationError, ValidationError, IdempotencyConflictError, PayloadTooLargeError):
                raise
        assert last_error is not None
        raise last_error

    # ---- High-level wire helpers used by the SDK ----

    def post_events(self, run_id: str, events: Sequence[Mapping[str, Any]], idempotency_key: Optional[str] = None) -> Dict[str, Any]:
        try:
            return self.request(
                "POST",
                f"/v1/runs/{run_id}/events",
                json_body={"events": list(events)},
                idempotency_key=idempotency_key,
            )
        except (CollectorUnavailableError, httpx.HTTPError) as exc:
            if self.spool is not None:
                self.spool.write({"run_id": run_id, "events": list(events), "idempotency_key": idempotency_key})
                return {"spooled": True, "reason": str(exc)}
            if self.fail_open:
                return {"spooled": False, "fail_open": True, "reason": str(exc)}
            raise

    def start_run(self, payload: Mapping[str, Any]) -> Dict[str, Any]:
        return self.request("POST", "/v1/runs/start", json_body=payload)

    def commit_run(self, run_id: str) -> Dict[str, Any]:
        return self.request("POST", f"/v1/runs/{run_id}/commit")

    def rollback_run(self, run_id: str, reason: Optional[str] = None) -> Dict[str, Any]:
        return self.request("POST", f"/v1/runs/{run_id}/rollback", json_body={"reason": reason})

    def policy_check(self, body: Mapping[str, Any]) -> Dict[str, Any]:
        return self.request("POST", "/v1/policy/check", json_body=body)

    def request_approval(self, body: Mapping[str, Any]) -> Dict[str, Any]:
        return self.request("POST", "/v1/approvals/request", json_body=body)

    def get_timeline(self, run_id: str) -> Dict[str, Any]:
        return self.request("GET", f"/v1/runs/{run_id}/timeline", retry=False)

    def verify_run(self, run_id: str) -> Dict[str, Any]:
        return self.request("GET", f"/v1/runs/{run_id}/verify", retry=False)

    def flush_spool(self) -> int:
        if not self.spool:
            return 0
        drained = self.spool.drain()
        flushed = 0
        for entry in drained:
            try:
                self.request(
                    "POST",
                    f"/v1/runs/{entry['run_id']}/events",
                    json_body={"events": entry["events"]},
                    idempotency_key=entry.get("idempotency_key"),
                )
                flushed += 1
            except Exception:
                # Re-queue and stop draining on first error.
                self.spool.write(entry)
                break
        return flushed

File: sdk/typescript/src/transport.ts
/**
 * HTTP transport with retry, batching, and an offline spool.
 *
 * Uses Node 18+ built-in fetch. No external HTTP dependencies.
 * Mirrors sdk/python/controldb/transport.py exactly.
 */

import { createWriteStream, existsSync, mkdirSync, readdirSync, readFileSync, unlinkSync } from "node:fs";
import { join } from "node:path";

import {
  AuthenticationError,
  AuthorizationError,
  CollectorUnavailableError,
  IdempotencyConflictError,
  PayloadTooLargeError,
  ValidationError,
} from "./errors.js";
import type { EventEnvelope } from "./events.js";

export interface TransportOptions {
  maxRetries?: number;
  /** Initial backoff delay in milliseconds (default: 200) */
  baseDelay?: number;
  /** Maximum backoff delay in milliseconds (default: 5000) */
  maxDelay?: number;
  timeout?: number;
  failOpen?: boolean;
  spoolPath?: string;
}

// ---- Spool ---------------------------------------------------------------

/**
 * Append-only on-disk spool for offline buffering.
 * Each call to write() appends one JSON line to a shared spool file.
 */
export class Spool {
  private readonly filePath: string;

  constructor(spoolPath: string) {
    mkdirSync(spoolPath, { recursive: true });
    this.filePath = join(spoolPath, "spool.ndjson");
  }

  write(entry: Record<string, unknown>): void {
    try {
      const ws = createWriteStream(this.filePath, { flags: "a", encoding: "utf8" });
      ws.write(JSON.stringify(entry) + "\n");
      ws.end();
    } catch {
      // Best-effort: ignore spool write errors
    }
  }

  drain(): Array<Record<string, unknown>> {
    if (!existsSync(this.filePath)) return [];
    let content: string;
    try {
      content = readFileSync(this.filePath, "utf8");
      unlinkSync(this.filePath);
    } catch {
      return [];
    }
    const entries: Array<Record<string, unknown>> = [];
    for (const line of content.split("\n")) {
      const trimmed = line.trim();
      if (!trimmed) continue;
      try {
        entries.push(JSON.parse(trimmed) as Record<string, unknown>);
      } catch {
        // Skip malformed lines
      }
    }
    return entries;
  }
}

// ---- Transport -----------------------------------------------------------

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function isRetryableStatus(status: number): boolean {
  return status >= 500;
}

async function raiseForStatus(resp: Response): Promise<void> {
  if (resp.status < 400) return;

  let message: string;
  try {
    const body = (await resp.json()) as Record<string, unknown>;
    message = (typeof body["detail"] === "string" ? body["detail"] : null) ?? resp.statusText;
  } catch {
    message = resp.statusText || String(resp.status);
  }

  switch (resp.status) {
    case 401:
      throw new AuthenticationError(message);
    case 403:
      throw new AuthorizationError(message);
    case 409:
      throw new IdempotencyConflictError(message);
    case 413:
      throw new PayloadTooLargeError(message);
    case 422:
      throw new ValidationError(message);
    default:
      throw new CollectorUnavailableError(message || `collector status ${resp.status}`);
  }
}

export class Transport {
  private readonly baseUrl: string;
  private readonly apiKey: string;
  private readonly maxRetries: number;
  private readonly baseDelay: number;
  private readonly maxDelay: number;
  private readonly failOpen: boolean;
  private readonly spool: Spool | null;

  constructor(baseUrl: string, apiKey: string, options: TransportOptions = {}) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.apiKey = apiKey;
    this.maxRetries = options.maxRetries ?? 4;
    this.baseDelay = options.baseDelay ?? 200;
    this.maxDelay = options.maxDelay ?? 5000;
    this.failOpen = options.failOpen ?? false;
    this.spool = options.spoolPath ? new Spool(options.spoolPath) : null;
  }

  async request(
    method: string,
    path: string,
    options: {
      body?: unknown;
      idempotencyKey?: string;
      retry?: boolean;
    } = {}
  ): Promise<Record<string, unknown>> {
    const { body, idempotencyKey, retry = true } = options;
    const attempts = retry ? this.maxRetries : 1;
    let lastError: unknown;

    for (let attempt = 0; attempt < attempts; attempt++) {
      try {
        const headers: Record<string, string> = {
          "Authorization": `Bearer ${this.apiKey}`,
          "Content-Type": "application/json",
          "User-Agent": "controldb-typescript/0.1",
        };
        if (idempotencyKey) {
          headers["Idempotency-Key"] = idempotencyKey;
        }

        const resp = await fetch(`${this.baseUrl}${path}`, {
          method,
          headers,
          body: body !== undefined ? JSON.stringify(body) : undefined,
        });

        // Retry on 5xx before throwing
        if (isRetryableStatus(resp.status) && attempt + 1 < attempts) {
          lastError = new CollectorUnavailableError(`collector status ${resp.status}`);
          const delay = Math.min(this.maxDelay, this.baseDelay * Math.pow(2, attempt));
          const jitter = Math.random() * (delay / 2);
          await sleep(delay + jitter);
          continue;
        }

        await raiseForStatus(resp);

        if (resp.status === 204 || resp.headers.get("content-length") === "0") {
          return {};
        }
        try {
          return (await resp.json()) as Record<string, unknown>;
        } catch {
          return {};
        }
      } catch (err) {
        // Non-retryable errors (auth, validation, etc.) propagate immediately
        if (
          err instanceof AuthenticationError ||
          err instanceof AuthorizationError ||
          err instanceof ValidationError ||
          err instanceof IdempotencyConflictError ||
          err instanceof PayloadTooLargeError
        ) {
          throw err;
        }
        lastError = err;
        if (attempt + 1 < attempts) {
          const delay = Math.min(this.maxDelay, this.baseDelay * Math.pow(2, attempt));
          const jitter = Math.random() * (delay / 2);
          await sleep(delay + jitter);
        }
      }
    }

    throw lastError instanceof Error
      ? lastError
      : new CollectorUnavailableError(String(lastError));
  }

  async postEvents(
    runId: string,
    events: EventEnvelope[],
    idempotencyKey?: string
  ): Promise<Record<string, unknown>> {
    try {
      return await this.request("POST", `/v1/runs/${runId}/events`, {
        body: { events },
        idempotencyKey,
      });
    } catch (err) {
      if (this.spool) {
        this.spool.write({ run_id: runId, events, idempotency_key: idempotencyKey ?? null });
        return { spooled: true, reason: String(err) };
      }
      if (this.failOpen) {
        return { spooled: false, fail_open: true, reason: String(err) };
      }
      throw err;
    }
  }

  async startRun(payload: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.request("POST", "/v1/runs/start", { body: payload });
  }

  async commitRun(runId: string): Promise<Record<string, unknown>> {
    return this.request("POST", `/v1/runs/${runId}/commit`);
  }

  async rollbackRun(runId: string, reason?: string): Promise<Record<string, unknown>> {
    return this.request("POST", `/v1/runs/${runId}/rollback`, {
      body: { reason: reason ?? null },
    });
  }

  async policyCheck(body: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.request("POST", "/v1/policy/check", { body });
  }

  async requestApproval(body: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.request("POST", "/v1/approvals/request", { body });
  }

  async getTimeline(runId: string): Promise<Record<string, unknown>> {
    return this.request("GET", `/v1/runs/${runId}/timeline`, { retry: false });
  }

  async verifyRun(runId: string): Promise<Record<string, unknown>> {
    return this.request("GET", `/v1/runs/${runId}/verify`, { retry: false });
  }

  async flushSpool(): Promise<number> {
    if (!this.spool) return 0;
    const drained = this.spool.drain();
    let flushed = 0;
    for (const entry of drained) {
      try {
        await this.request("POST", `/v1/runs/${entry["run_id"]}/events`, {
          body: { events: entry["events"] },
          idempotencyKey: (entry["idempotency_key"] as string | undefined) ?? undefined,
        });
        flushed++;
      } catch {
        // Re-queue and stop draining on first error
        this.spool.write(entry);
        break;
      }
    }
    return flushed;
  }

  async close(): Promise<void> {
    // No persistent connections to close with fetch; included for API parity
  }
}
